src/socket/icmpv4.rs

```diff
@@ -0,0 +1,378 @@
+use core::cmp;
+use managed::Managed;
+
+use {Error, Result};
+use phy::DeviceCapabilities;
+use socket::{Socket, SocketHandle};
+use storage::{Resettable, RingBuffer};
+use wire::{IpProtocol, Ipv4Address, Ipv4Repr, Icmpv4Repr, Icmpv4Packet};
+
+/// A buffered ICMPv4 packet.
```

Comment:
I think that ICMP sockets could easily serve both IPv4 and IPv6. After all, the first nibble of the header already tells you which packet it is.

Reply:
True

Reply:
Id be interested to see how the type based `binds` are implemented for IPv6. AFAIK we'd need to support the Fragment header to get an identification number, but the id is 32 bits in IPv6. I'll have to dig through linux and/or some docs.

Reply:
Don't bother with IPv6 for now, its headers are fairly broken and we'll need to aggressively compress them into Reprs.